File: src/macos/LanMessenger/Core/Networking/NetworkCoordinator.swift
import Foundation

// Owns all PeerSessions, the DiscoveryService, and the NetworkInterfaceMonitor.
// Routes validated packets to MessagingService / FileTransferService via the
// delegate. Manages the TCP listener for inbound connections from peers.
//
// All delegate callbacks fire on the main queue.

protocol NetworkCoordinatorDelegate: AnyObject {
    @MainActor func coordinator(_ c: NetworkCoordinator, didReceivePacket packet: ValidatedPacket)
    @MainActor func coordinator(_ c: NetworkCoordinator, didDiscoverPeer packet: DiscoveryPacket, fromIP: String)
    @MainActor func coordinatorNetworkAvailabilityChanged(_ c: NetworkCoordinator)
}

// Default no-op for the availability hook so existing implementers don't need to add it.
extension NetworkCoordinatorDelegate {
    @MainActor func coordinatorNetworkAvailabilityChanged(_ c: NetworkCoordinator) {}
}

final class NetworkCoordinator: NSObject {

    weak var delegate: NetworkCoordinatorDelegate?

    let network: NetworkInterfaceMonitor
    let discovery: DiscoveryService

    private var sessions: [String: PeerSession] = [:]   // keyed by peerIP
    private var listenerSocket: Int32 = -1
    private let listenerQueue = DispatchQueue(label: "com.dave.lanmessenger.listener", qos: .utility)
    private let tcpPort: UInt16 = 54232
    private var running = false
    private var monitorObserverID: UUID?

    var isLocalNetworkAvailable: Bool { network.isLocalNetworkAvailable }
    var localIPs: Set<String> { network.localIPs }

    private var ownPublicKeyB64: String { KeyManager.shared.publicKeyB64 }

    override init() {
        let monitor = NetworkInterfaceMonitor()
        self.network = monitor
        self.discovery = DiscoveryService(monitor: monitor)
        super.init()
    }

    func start() {
        guard !running else { return }
        running = true

        // Monitor must be running BEFORE discovery so the first beacon sees a
        // populated interface set.
        network.start()
        monitorObserverID = network.addObserver { [weak self] in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.delegate?.coordinatorNetworkAvailabilityChanged(self)
            }
        }

        discovery.ownPublicKeyB64 = ownPublicKeyB64
        discovery.delegate = self
        // Read the username fresh on every beacon so changes in Settings
        // propagate without restarting the network stack.
        discovery.buildPayload = { [weak self] in
            DiscoveryPacket(
                type: "discovery",
                username: ConfigStore.shared.config.username,
                port: 54232,
                publicKeyB64: self?.ownPublicKeyB64 ?? "",
                ips: self?.network.localIPs.map { $0 } ?? []
            )
        }
        discovery.extraTargets = { [weak self] in
            self?.sessions.keys.map { $0 } ?? []
        }
        discovery.start()

        startTCPListener()

        NetLogger.info("Net",
            "coordinator started user='\(ConfigStore.shared.config.username)' tcp=\(tcpPort) udp=54231 " +
            "interfaces=\(network.adapters.count) available=\(network.isLocalNetworkAvailable)")
    }

    func stop() {
        running = false
        discovery.stop()
        if let id = monitorObserverID { network.removeObserver(id); monitorObserverID = nil }
        network.stop()
        sessions.values.forEach { $0.stop() }
        sessions.removeAll()
        if listenerSocket >= 0 { Darwin.close(listenerSocket); listenerSocket = -1 }
    }

    // Send a pre-encoded frame to a peer by IP. Opens a one-shot connection if needed.
    func send(frame: Data, toIP: String, port: Int = 54232) {
        if let session = sessions[toIP] {
            session.send(frame)
        } else {
            DispatchQueue.global(qos: .utility).async {
                guard let socket = self.openSocket(ip: toIP, port: port) else { return }
                defer { Darwin.close(socket) }
                _ = frame.withUnsafeBytes { ptr in
                    Darwin.send(socket, ptr.baseAddress, frame.count, 0)
                }
            }
        }
    }

    func send(frames: [Data], toIP: String, port: Int = 54232) {
        for frame in frames { send(frame: frame, toIP: toIP, port: port) }
    }

    // Ensure a persistent session exists for this peer.
    func ensureSession(ip: String, port: Int) {
        guard sessions[ip] == nil else { return }
        let session = PeerSession(ip: ip, port: port)
        session.ownPublicKeyB64 = ownPublicKeyB64
        session.onPacket = { [weak self] pkt in
            guard let self else { return }
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.delegate?.coordinator(self, didReceivePacket: pkt)
            }
        }
        session.onDisconnect = { [weak self] s in
            self?.sessions.removeValue(forKey: s.peerIP)
        }
        sessions[ip] = session
        session.start()
    }

    // MARK: - TCP Listener (inbound connections)

    private func startTCPListener() {
        listenerSocket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)
        guard listenerSocket >= 0 else {
            NetLogger.error("Net", "listener socket() failed: \(String(cString: strerror(errno)))")
            return
        }

        var enable: Int32 = 1
        setsockopt(listenerSocket, SOL_SOCKET, SO_REUSEADDR, &enable, socklen_t(MemoryLayout<Int32>.size))

        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = tcpPort.bigEndian
        addr.sin_addr.s_addr = INADDR_ANY
        let bindResult = withUnsafePointer(to: &addr) { ptr -> Int32 in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                Darwin.bind(listenerSocket, sockPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        if bindResult != 0 {
            NetLogger.error("Net", "listener bind() failed: \(String(cString: strerror(errno)))")
        }
        _ = listen(listenerSocket, 16)
        NetLogger.info("Net", "TCP listener bound on 0.0.0.0:\(tcpPort)")

        listenerQueue.async { [weak self] in
            while self?.running == true {
                var clientAddr = sockaddr_in()
                var addrLen = socklen_t(MemoryLayout<sockaddr_in>.size)
                let clientSocket = withUnsafeMutablePointer(to: &clientAddr) { ptr in
                    ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                        accept(self?.listenerSocket ?? -1, sockPtr, &addrLen)
                    }
                }
                guard clientSocket >= 0 else { continue }
                let ip = String(cString: inet_ntoa(clientAddr.sin_addr))
                self?.handleInbound(socket: clientSocket, fromIP: ip)
            }
        }
    }

    private func handleInbound(socket: Int32, fromIP: String) {
        DispatchQueue.global(qos: .utility).async { [weak self] in
            defer { Darwin.close(socket) }

            // 30 s read timeout — kills stuck readers without losing fresh data
            var tv = timeval(tv_sec: 30, tv_usec: 0)
            setsockopt(socket, SOL_SOCKET, SO_RCVTIMEO, &tv, socklen_t(MemoryLayout<timeval>.size))

            while true {
                guard let frameData = Self.recvFrame(socket: socket) else { break }
                guard let json = try? JSONSerialization.jsonObject(with: frameData) as? [String: Any] else { break }
                let result = PacketValidator.validate(
                    json: json,
                    senderIP: fromIP,
                    ownPublicKeyB64: self?.ownPublicKeyB64 ?? ""
                )
                if case .success(let pkt) = result {
                    DispatchQueue.main.async { [weak self] in
                        guard let self else { return }
                        self.delegate?.coordinator(self, didReceivePacket: pkt)
                    }
                }
            }
        }
    }

    // Reads one length-prefixed frame directly from the socket using recv().
    // CFStream wrappers around accepted sockets have been unreliable in practice;
    // raw recv() with a socket-level read timeout works deterministically.
    private static func recvFrame(socket: Int32) -> Data? {
        var header = [UInt8](repeating: 0, count: 4)
        guard recvExact(socket: socket, buffer: &header, count: 4) else { return nil }
        let length = Int(UInt32(bigEndian: header.withUnsafeBytes { $0.load(as: UInt32.self) }))
        guard length > 0, length <= FrameCodec.maxFrameSize else { return nil }
        var body = [UInt8](repeating: 0, count: length)
        guard recvExact(socket: socket, buffer: &body, count: length) else { return nil }
        return Data(body)
    }

    private static func recvExact(socket: Int32, buffer: inout [UInt8], count: Int) -> Bool {
        var total = 0
        while total < count {
            let n = buffer.withUnsafeMutableBytes { ptr in
                Darwin.recv(socket, ptr.baseAddress!.advanced(by: total), count - total, 0)
            }
            if n <= 0 { return false }
            total += n
        }
        return true
    }

    // MARK: - Helpers

    private func openSocket(ip: String, port: Int) -> Int32? {
        let fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)
        guard fd >= 0 else { return nil }
        var addr = sockaddr_in()
        addr.sin_family = sa_family_t(AF_INET)
        addr.sin_port = UInt16(port).bigEndian
        addr.sin_addr.s_addr = inet_addr(ip)
        let result = withUnsafePointer(to: &addr) { ptr in
            ptr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockPtr in
                Darwin.connect(fd, sockPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
        }
        if result != 0 { Darwin.close(fd); return nil }
        return fd
    }
}

// MARK: - DiscoveryServiceDelegate

extension NetworkCoordinator: DiscoveryServiceDelegate {
    func discoveryService(_ service: DiscoveryService, didDiscoverPeer packet: DiscoveryPacket, fromIP: String) {
        Task { @MainActor [weak self] in
            guard let self else { return }
            self.delegate?.coordinator(self, didDiscoverPeer: packet, fromIP: fromIP)
        }
    }
}

File: src/windows-native/LanMessenger/Core/Networking/NetworkCoordinator.cs
using LanMessenger.Core.Crypto;
using LanMessenger.Core.Protocol;
using LanMessenger.Core.Services;
using Microsoft.UI.Dispatching;
using System.Net;
using System.Net.Sockets;

namespace LanMessenger.Core.Networking;

// Owns all PeerSessions, the DiscoveryService, and the NetworkInterfaceMonitor.
// Routes validated packets to callbacks (MessagingService / FileTransferService).
// Manages the TCP listener for inbound connections from peers.
// Callbacks are dispatched to the UI thread via the stored DispatcherQueue.
public sealed class NetworkCoordinator : IDisposable
{
    public event Action<ValidatedPacket>?         PacketReceived;
    public event Action<DiscoveryPacket, string>? PeerDiscovered;
    public event Action<string, string>?          PeerDeparted;   // (publicKeyB64, fromIP)
    public event Action?                          NetworkAvailabilityChanged;

    // Extra unicast beacon targets (last-known IPs of saved contacts that are
    // not currently online). Supplied by AppModel; lets discovery reach peers
    // that broadcast/multicast can't — different subnets, APs with broadcast
    // isolation, or a peer whose beacons are firewalled while unicast works.
    // Invoked on the beacon timer thread; DiscoveryService guards the call.
    public Func<IEnumerable<string>>? UnicastHints { get; set; }

    public readonly NetworkInterfaceMonitor Network  = new();
    public readonly DiscoveryService        Discovery;
    private readonly Dictionary<string, PeerSession> _sessions = [];
    private TcpListener? _listener;
    private CancellationTokenSource _cts = new();
    private bool _running;
    private DispatcherQueue? _dispatcherQueue;

    private const int TcpPort = 54232;

    public bool IsLocalNetworkAvailable => Network.IsLocalNetworkAvailable;

    private string OwnPublicKeyB64 => KeyManager.Shared.PublicKeyB64;

    public NetworkCoordinator()
    {
        Discovery = new DiscoveryService(Network);
    }

    public void Start(DispatcherQueue dispatcherQueue)
    {
        if (_running) return;
        _running = true;
        _cts = new CancellationTokenSource();
        _dispatcherQueue = dispatcherQueue;

        // Monitor must be running BEFORE Discovery so the first beacon sees a
        // populated interface set.
        Network.Start();
        Network.Changed += () =>
            _dispatcherQueue?.TryEnqueue(() => NetworkAvailabilityChanged?.Invoke());

        // Configure discovery — read the username fresh on every beacon so
        // changes the user makes in Settings propagate without needing to
        // restart the network stack.
        Discovery.OwnPublicKeyB64 = OwnPublicKeyB64;
        Discovery.BuildPayload    = () => new DiscoveryPacket
        {
            Type         = "discovery",
            Username     = LanMessenger.Core.Persistence.ConfigStore.Shared.Config.Username,
            Port         = TcpPort,
            PublicKeyB64 = OwnPublicKeyB64,
            Ips          = [.. Network.LocalIPs],
            // relay_id_hash lets peers know where to send cloud-relay messages
            // destined for us. It is SHA256(relay_id) and safe to publish.
            RelayIdHash  = LanMessenger.Core.Services.RelayClient.Shared.RelayIdHash(),
        };
        Discovery.ExtraTargets = () => UnicastHints?.Invoke() ?? [];
        Discovery.PeerDiscovered += (pkt, ip) =>
            _dispatcherQueue?.TryEnqueue(() => PeerDiscovered?.Invoke(pkt, ip));
        Discovery.PeerDeparted += (key, ip) =>
            _dispatcherQueue?.TryEnqueue(() => PeerDeparted?.Invoke(key, ip));
        Discovery.Start();

        StartTcpListener();

        LanLogger.Info("Net",
            $"coordinator started user='{LanMessenger.Core.Persistence.ConfigStore.Shared.Config.Username}' tcp={TcpPort} udp=54231 " +
            $"interfaces={Network.Adapters.Count} available={Network.IsLocalNetworkAvailable}");
    }

    public void Stop()
    {
        _running = false;
        _cts.Cancel();
        Discovery.Stop();
        Network.Stop();
        lock (_sessions)
        {
            foreach (var s in _sessions.Values) s.Stop();
            _sessions.Clear();
        }
        try { _listener?.Stop(); } catch { }
        _listener = null;
    }

    public void Dispose() => Stop();

    // Send a pre-encoded frame to a peer. Uses persistent session if available,
    // else fire-and-forget TCP.
    public void Send(byte[] frame, string toIP, int port = TcpPort)
    {
        PeerSession? session;
        lock (_sessions) _sessions.TryGetValue(toIP, out session);

        if (session is not null)
        {
            session.Send(frame);
            return;
        }

        Task.Run(async () =>
        {
            try
            {
                using var tcp = new TcpClient();
                await tcp.ConnectAsync(toIP, port).WaitAsync(TimeSpan.FromSeconds(5)).ConfigureAwait(false);
                await tcp.GetStream().WriteAsync(frame).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                LanLogger.Info("Net", $"one-shot send to {toIP}:{port} failed: {ex.GetType().Name} {ex.Message}");
            }
        });
    }

    public void Send(IEnumerable<byte[]> frames, string toIP, int port = TcpPort)
    {
        foreach (var f in frames) Send(f, toIP, port);
    }

    // Announce our departure so peers flip us offline immediately.
    public void SendGoodbye() => Discovery.SendGoodbye();

    // Actively reconfirm a quiet peer before declaring it offline.
    public void Probe(string ip) => Discovery.Probe(ip);

    // Ensure a persistent session exists for this peer.
    public void EnsureSession(string ip, int port)
    {
        lock (_sessions)
        {
            if (_sessions.ContainsKey(ip)) return;
            var session = new PeerSession(ip, port) { OwnPublicKeyB64 = OwnPublicKeyB64 };
            session.OnPacket += pkt =>
                _dispatcherQueue?.TryEnqueue(() => PacketReceived?.Invoke(pkt));
            session.OnDisconnect += s => { lock (_sessions) _sessions.Remove(s.PeerIP); };
            _sessions[ip] = session;
            session.Start();
        }
    }

    // MARK: - TCP Listener (inbound connections)

    private void StartTcpListener()
    {
        _listener = new TcpListener(IPAddress.Any, TcpPort);
        _listener.Server.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
        _listener.Start(backlog: 16); // throws SocketException if port is already bound
        Task.Run(() => AcceptLoop(_cts.Token));
        LanLogger.Info("Net", $"TCP listener bound on 0.0.0.0:{TcpPort}");
    }

    private async Task AcceptLoop(CancellationToken ct)
    {
        var consecutiveFailures = 0;
        while (!ct.IsCancellationRequested && _running)
        {
            try
            {
                var client = await _listener!.AcceptTcpClientAsync(ct).ConfigureAwait(false);
                consecutiveFailures = 0;
                var fromIP = ((IPEndPoint)client.Client.RemoteEndPoint!).Address.ToString();
                _ = Task.Run(() => HandleInbound(client, fromIP, ct));
            }
            catch (OperationCanceledException) { break; }
            catch (Exception ex)
            {
                LanLogger.Info("Net", $"accept failed: {ex.GetType().Name} {ex.Message}");
                // A dead listener (driver reset, socket error) would otherwise
                // spin this loop at 100% CPU and silently stop all inbound
                // messaging — the "can receive nothing until app restart"
                // failure mode. Back off, then try rebuilding the listener.
                consecutiveFailures++;
                try { await Task.Delay(Math.Min(consecutiveFailures, 10) * 500, ct).ConfigureAwait(false); }
                catch (OperationCanceledException) { break; }
                if (consecutiveFailures >= 3 && _running)
                {
                    try
                    {
                        try { _listener?.Stop(); } catch { }
                        _listener = new TcpListener(IPAddress.Any, TcpPort);
                        _listener.Server.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
                        _listener.Start(backlog: 16);
                        consecutiveFailures = 0;
                        LanLogger.Info("Net", $"TCP listener rebuilt on 0.0.0.0:{TcpPort}");
                    }
                    catch (Exception rebuildEx)
                    {
                        LanLogger.Warn("Net", $"TCP listener rebuild failed: {rebuildEx.GetType().Name} {rebuildEx.Message}");
                    }
                }
            }
        }
    }

    private async Task HandleInbound(TcpClient client, string fromIP, CancellationToken ct)
    {
        using (client)
        {
            try
            {
                // Per-frame idle timeout so half-open / abandoned connections are
                // released promptly. (Socket.ReceiveTimeout only applies to
                // synchronous reads — it silently does nothing for the async
                // reads used here, which previously left crashed-peer sockets
                // alive indefinitely.)
                var stream = client.GetStream();
                while (!ct.IsCancellationRequested)
                {
                    using var idleCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
                    idleCts.CancelAfter(TimeSpan.FromSeconds(60));
                    byte[]? frameData;
                    try
                    {
                        frameData = await FrameCodec.ReadFrameAsync(stream, idleCts.Token).ConfigureAwait(false);
                    }
                    catch (OperationCanceledException) when (!ct.IsCancellationRequested)
                    {
                        LanLogger.Info("Net", $"inbound from {fromIP} idle-timed out");
                        break;
                    }
                    if (frameData is null) break;

                    var pkt = PacketValidator.Validate(frameData, fromIP, OwnPublicKeyB64);
                    if (pkt is not null)
                        _dispatcherQueue?.TryEnqueue(() => PacketReceived?.Invoke(pkt));
                    else
                        LanLogger.Warn("Net", $"dropped invalid frame from {fromIP} bytes={frameData.Length}");
                }
            }
            catch (Exception ex) { LanLogger.Warn("Net", $"inbound from {fromIP} ended: {ex.GetType().Name} {ex.Message}"); }
        }
    }
}
